src/file/format/hjson.rs

```diff
@@ -1,17 +1,41 @@
 use std::collections::HashMap;
 use std::error::Error;
 
-use serde_hjson;
-
+use crate::error::{ConfigError, Unexpected};
 use crate::value::{Value, ValueKind};
 
 pub fn parse(
     uri: Option<&String>,
     text: &str,
 ) -> Result<HashMap<String, Value>, Box<dyn Error + Send + Sync>> {
     // Parse a JSON object value from the text
-    // TODO: Have a proper error fire if the root of a file is ever not a Table
-    let value = from_hjson_value(uri, &serde_hjson::from_str(text)?);
+    let root = serde_hjson::from_str(text)?;
+    match root {
+        serde_hjson::Value::String(value) => {
+            return Err(ConfigError::invalid_root(uri, Unexpected::Str(value)))
+        }
+        serde_hjson::Value::I64(value) => {
+            return Err(ConfigError::invalid_root(uri, Unexpected::Integer(value)))
+        }
+        serde_hjson::Value::U64(value) => {
+            return Err(ConfigError::invalid_root(
+                uri,
+                Unexpected::Integer(value as i64),
+            ))
+        }
+        serde_hjson::Value::F64(value) => {
+            return Err(ConfigError::invalid_root(uri, Unexpected::Float(value)))
+        }
+        serde_hjson::Value::Bool(value) => {
+            return Err(ConfigError::invalid_root(uri, Unexpected::Bool(value)))
+        }
+        serde_hjson::Value::Object(_) => (),
+        serde_hjson::Value::Array(_) => {
+            return Err(ConfigError::invalid_root(uri, Unexpected::Seq))
+        }
+        serde_hjson::Value::Null => return Err(ConfigError::invalid_root(uri, Unexpected::Unit)),
+    };
```

Comment:
I guess we can simplify this whole `match` to something like this:

```
let unexp_err = match {
    serde_hjson::Value::Object(_) => None,
    serde_hjson::Value::String(v) => Unexpected::Str(value),
    /* and the other Unexpected instances */
};

if let Some(unex) = unexp_err {
    return Err(ConfigError::unvalid_root(uri, unex))
}
```

which would reduce the mental load on this one a lot, IMO. :smile: